Sources/AsyncAlgorithms/Channels/ChannelStorage.swift

```diff
@@ -0,0 +1,152 @@
+  var onSendSuspended: (() -> Void)?
+  var onNextSuspended: (() -> Void)?
+#endif
+
+  init() {
+    self.stateMachine = ManagedCriticalState(ChannelStateMachine())
+  }
+
+  func generateId() -> Int {
+    self.ids.withCriticalRegion { ids in
+      defer { ids += 1 }
+      return ids
+    }
+  }
+
+  func send(element: Element) async {
+    let producerId = self.generateId()
+    let wasCancelled = ManagedCriticalState(false)
+
+    await withTaskCancellationHandler {
+      await withUnsafeContinuation { (continuation: UnsafeContinuation<Void, Never>) in
+        self.stateMachine.withCriticalRegion { stateMachine in
+          let wasCancelled = wasCancelled.withCriticalRegion { $0 }
+          guard !wasCancelled else {
+            continuation.resume()
+            return
+          }
+
+          let action = stateMachine.newProducer(continuation: continuation, element: element, producerId: producerId)
+
+          switch action {
+            case .suspend:
+#if DEBUG
+              self.onSendSuspended?()
+#endif
+              break
+            case .resumeProducer:
+              continuation.resume()
+            case .resumeProducerAndConsumer(let consumerContinuation):
+              continuation.resume()
+              consumerContinuation?.resume(returning: element)
+          }
+        }
+      }
+    } onCancel: {
+      self.stateMachine.withCriticalRegion { stateMachine in
+        wasCancelled.withCriticalRegion { $0 = true }
+        let action = stateMachine.producerWasCancelled(producerId: producerId)
+
+        switch action {
+          case .none:
+            break
+          case .resumeProducer(let continuation):
+            continuation?.resume()
+        }
+      }
+    }
+  }
+
+  func fail(error: Error) where Failure == Error {
+    self.stateMachine.withCriticalRegion { stateMachine in
+      let action = stateMachine.fail(error: error)
+
+      switch action {
+        case .none:
+          break
+        case .resumeProducers(let continuations):
+          continuations.forEach { $0?.resume() }
+        case .resumeConsumers(let continuations):
+          continuations.forEach { $0?.resume(throwing: error) }
+      }
+    }
+  }
+
+  func finish() {
+    self.stateMachine.withCriticalRegion { stateMachine in
+      let action = stateMachine.finish()
+
+      switch action {
+        case .none:
+          break
+        case .resumeProducers(let continuations):
+          continuations.forEach { $0?.resume() }
+        case .resumeConsumers(let continuations):
+          continuations.forEach { $0?.resume(returning: nil) }
+      }
+    }
+  }
+
+  func next() async throws -> Element? {
+    guard !Task.isCancelled else { return nil }
+
+    let consumerId = self.generateId()
+    let wasCancelled = ManagedCriticalState(false)
+
+    return try await withTaskCancellationHandler {
+      return try await withUnsafeThrowingContinuation { (continuation: UnsafeContinuation<Element?, any Error>) in
```

Comment:
Same comments as above. I don't think we should track `wasCancelled` here. Furthermore, we should only conditionally create a continuation.

Reply:
See my comment about tracking the already cancelled send/next thanks to `Set`s in the channeling state.